lib/spack/spack/directives.py

```diff
@@ -180,24 +180,43 @@ def _depends_on(pkg, spec, when=None):
         when = pkg.name
     when_spec = parse_anonymous_spec(when, pkg.name)
 
+    if deptypes is None:
+        # The default deptype is build and link because the common case is to
+        # build against a library which then turns into a runtime dependency
+        # due to the linker.
```

Comment:
Yep.

To understand this you have to distinguish runtime from build time.  `run` dependencies are for when a user runs the installed package.  `build` dependencies are for when *spack* builds the package.  Those are two different environments.  This is the same reason why things like `Package.setup_environment()` distinguish between `spack_env` and `run_env`.